File: src/android/AppOpenAdExecutor.java
package com.emi.cordova.admob.nextgen;

import static com.emi.cordova.admob.nextgen.AdMobNextGen.isInitialized;

import android.app.Activity;
import android.util.Log;

import androidx.annotation.NonNull;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Date;

import com.google.android.libraries.ads.mobile.sdk.appopen.AppOpenAd;
import com.google.android.libraries.ads.mobile.sdk.appopen.AppOpenAdEventCallback;
import com.google.android.libraries.ads.mobile.sdk.common.AdLoadCallback;
import com.google.android.libraries.ads.mobile.sdk.common.AdRequest;
import com.google.android.libraries.ads.mobile.sdk.common.AdValue;
import com.google.android.libraries.ads.mobile.sdk.common.FullScreenContentError;
import com.google.android.libraries.ads.mobile.sdk.common.LoadAdError;

public class AppOpenAdExecutor {

    private static final String TAG = "AdMobAppOpen";
    private static AppOpenAdExecutor instance;

    private CordovaInterface cordova;
    private CordovaWebView webView;

    private AppOpenAd appOpenAd;

    private boolean isLoadingAd = false;
    private boolean isShowingAd = false;
    private long loadTime = 0;

    private boolean isAutoShow = false;
    private String currentAdUnitId = null;

    private long lastLoadTime = 0;
    private long minLoadInterval = 5000;

    public static synchronized AppOpenAdExecutor getInstance() {
        if (instance == null) {
            instance = new AppOpenAdExecutor();
        }
        return instance;
    }

    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        this.cordova = cordova;
        this.webView = webView;
    }

    public void loadAd(JSONArray args, CallbackContext callbackContext) {
        if (!isInitialized) {
            callbackContext.error("SDK not initialized");
            return;
        }
        try {
            JSONObject options = args.optJSONObject(0);
            String adUnitId;

            if (options != null) {
                adUnitId = options.getString("adUnitId");
                if (options.has("isAutoShow")) this.isAutoShow = options.getBoolean("isAutoShow");
                if (options.has("retryInterval")) this.minLoadInterval = options.getLong("retryInterval");
            } else {
                adUnitId = args.getString(0);
                this.isAutoShow = false;
            }

            this.currentAdUnitId = adUnitId;
            loadAdInternal(adUnitId, callbackContext);

        } catch (JSONException e) {
            if (callbackContext != null) callbackContext.error("Invalid Args: " + e.getMessage());
        }
    }

    private void loadAdInternal(String adUnitId, CallbackContext callbackContext) {
        long currentTime = new Date().getTime();

        if (isLoadingAd || isAdAvailable()) {
            if (isAdAvailable()) {
                fireEvent("on.appopen.loaded", null);
                if (callbackContext != null) callbackContext.success("Ad Ready");
            }
            return;
        }

        if ((currentTime - lastLoadTime) < minLoadInterval) {
            if (callbackContext != null) callbackContext.error("Request too fast. Please wait " + minLoadInterval + " ms to prevent invalid traffic.");
            return;
        }

        isLoadingAd = true;
        lastLoadTime = currentTime;

        if (callbackContext != null) {
            PluginResult r = new PluginResult(PluginResult.Status.OK, "Loading started...");
            r.setKeepCallback(true);
            callbackContext.sendPluginResult(r);
        }

        cordova.getActivity().runOnUiThread(() -> {

            AdRequest request = new AdRequest.Builder(adUnitId).build();

            AppOpenAd.load(
                    request,
                    new AdLoadCallback<AppOpenAd>() {
                        @Override
                        public void onAdLoaded(@NonNull AppOpenAd ad) {
                            isLoadingAd = false;
                            appOpenAd = ad;
                            loadTime = new Date().getTime();

                            fireEvent("on.appopen.loaded", null);
                        }

                        @Override
                        public void onAdFailedToLoad(@NonNull LoadAdError loadAdError) {
                            isLoadingAd = false;

                            try {
                                JSONObject errData = new JSONObject();
                                errData.put("code", loadAdError.getCode());
                                errData.put("message", loadAdError.getMessage());
                                fireEvent("on.appopen.failed.load", errData);
                            } catch (JSONException e) {}
                        }
                    }
            );
        });
    }

    public void showAdIfAvailable(@NonNull Activity activity) {
        if (isShowingAd) return;

        if (!isAdAvailable()) {
            try {
                JSONObject errData = new JSONObject();
                errData.put("message", "Ad not ready or expired");
                fireEvent("on.appopen.failed.show", errData);
            } catch (JSONException e) {}

            if (currentAdUnitId != null) loadAdInternal(currentAdUnitId, null);
            return;
        }

        appOpenAd.setAdEventCallback(new AppOpenAdEventCallback() {

            @Override
            public void onAdPaid(@NonNull AdValue adValue) {
                try {
                    JSONObject data = new JSONObject();
                    data.put("value", adValue.getValueMicros());
                    data.put("currency", adValue.getCurrencyCode());
                    data.put("precision", adValue.getPrecisionType());
                    fireEvent("on.appopen.revenue", data);
                } catch (JSONException e) {}
            }

            @Override
            public void onAdShowedFullScreenContent() {
                isShowingAd = true;
                fireEvent("on.appopen.shown", null);
            }

            @Override
            public void onAdDismissedFullScreenContent() {
                appOpenAd = null;
                isShowingAd = false;
                fireEvent("on.appopen.dismissed", null);
            }

            @Override
            public void onAdFailedToShowFullScreenContent(@NonNull FullScreenContentError error) {
                appOpenAd = null;
                isShowingAd = false;
                try {
                    JSONObject errData = new JSONObject();
                    errData.put("code", error.getCode());
                    errData.put("message", error.getMessage());
                    fireEvent("on.appopen.failed.show", errData);
                } catch (JSONException e) {}
            }

            @Override
            public void onAdImpression() {
                fireEvent("on.appopen.impression", null);
            }

            @Override
            public void onAdClicked() {
                fireEvent("on.appopen.clicked", null);
            }
        });

        cordova.getActivity().runOnUiThread(() -> {

            if (!AdMobNextGen.isAppInForeground) {
                try {
                    JSONObject errData = new JSONObject();
                    errData.put("code", "APP_IN_BACKGROUND");
                    errData.put("message", "Ad presentation blocked: Application is currently in the background.");
                    fireEvent("on.appopen.failed.show", errData);
                } catch (JSONException e) {}

                return;
            }

            isShowingAd = true;
            appOpenAd.show(activity);
        });

    }

    public boolean shouldAutoShow() {
        return this.isAutoShow;
    }

    public void setAutoShow(boolean value) {
        this.isAutoShow = value;
    }

    private boolean wasLoadTimeLessThanNHoursAgo(long numHours) {
        long dateDifference = new Date().getTime() - loadTime;
        long numMilliSecondsPerHour = 3600000L;
        return dateDifference < (numMilliSecondsPerHour * numHours);
    }

    private boolean isAdAvailable() {
        return appOpenAd != null && wasLoadTimeLessThanNHoursAgo(4);
    }

    private void fireEvent(String eventName, JSONObject data) {
        if (cordova == null) return;
        cordova.getActivity().runOnUiThread(() -> {
            StringBuilder js = new StringBuilder();
            js.append("javascript:cordova.fireDocumentEvent('");
            js.append(eventName);
            js.append("'");
            if (data != null) {
                js.append(", ");
                js.append(data.toString());
            }
            js.append(");");
            if (webView != null && webView.getView() != null) webView.loadUrl(js.toString());
        });
    }
}

File: src/ios/AppOpenAdExecutor.m
#import "AppOpenAdExecutor.h"
#import "AdMobNextGen.h"

@interface AppOpenAdExecutor()

@property (nonatomic, weak) AdMobNextGen *plugin;
@property (nonatomic, strong) GADAppOpenAd *appOpenAd;

@property (nonatomic, assign) BOOL isLoadingAd;
@property (nonatomic, assign) BOOL isShowingAd;
@property (nonatomic, strong) NSDate *loadTime;

@property (nonatomic, assign) BOOL isAutoShow;
@property (nonatomic, strong) NSString *currentAdUnitId;

@property (nonatomic, assign) NSTimeInterval lastLoadTime;
@property (nonatomic, assign) NSTimeInterval minLoadInterval;

@end

@implementation AppOpenAdExecutor

+ (instancetype)sharedInstance {
    static AppOpenAdExecutor *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[AppOpenAdExecutor alloc] init];
    });
    return instance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        self.isLoadingAd = NO;
        self.isShowingAd = NO;
        self.isAutoShow = NO;
        self.lastLoadTime = 0;
        self.minLoadInterval = 5.0; 
    }
    return self;
}

- (void)initializeWithPlugin:(AdMobNextGen *)plugin {
    self.plugin = plugin;
}

- (void)loadAppOpenAd:(NSDictionary *)options command:(CDVInvokedUrlCommand *)command {
    NSString *adUnitId = options[@"adUnitId"];

    if (!adUnitId) {
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"adUnitId is required."];
        [self.plugin.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        return;
    }

    if (options[@"isAutoShow"] != nil) {
        self.isAutoShow = [options[@"isAutoShow"] boolValue];
    } else {
        self.isAutoShow = NO;
    }

    if (options[@"retryInterval"] != nil) {
        self.minLoadInterval = [options[@"retryInterval"] doubleValue] / 1000.0;
    }

    self.currentAdUnitId = adUnitId;
    [self loadAdInternal:adUnitId command:command];
}

- (void)loadAdInternal:(NSString *)adUnitId command:(CDVInvokedUrlCommand *)command {
    dispatch_async(dispatch_get_main_queue(), ^{
        NSTimeInterval currentTime = [[NSDate date] timeIntervalSince1970];

        if (self.isLoadingAd || [self isAdAvailable]) {
            if ([self isAdAvailable]) {
                [self.plugin fireEvent:@"document" event:@"on.appopen.loaded" withData:nil];
                if (command) {
                    CDVPluginResult* res = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"Ad Ready"];
                    [self.plugin.commandDelegate sendPluginResult:res callbackId:command.callbackId];
                }

                if (self.isAutoShow) {
                    [self showAdIfAvailable];
                }

            }
            return;
        }

        if ((currentTime - self.lastLoadTime) < self.minLoadInterval) {

            NSString *errorMsg = [NSString stringWithFormat:@"Request too fast. Please wait %.0fms to prevent invalid traffic.", (self.minLoadInterval * 1000.0)];

            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:errorMsg];
            [self.plugin.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
            return;
        }

        self.isLoadingAd = YES;
        self.lastLoadTime = currentTime;

        if (command) {
            CDVPluginResult* res = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"Loading started..."];
            [res setKeepCallbackAsBool:YES];
            [self.plugin.commandDelegate sendPluginResult:res callbackId:command.callbackId];
        }

        GADRequest *request = [GADRequest request];

        [GADAppOpenAd loadWithAdUnitID:adUnitId
                               request:request
                     completionHandler:^(GADAppOpenAd *ad, NSError *error) {
            self.isLoadingAd = NO;

            if (error) {

                NSString *jsonStr = [NSString stringWithFormat:@"{\"code\":%ld, \"message\":\"%@\"}", (long)error.code, error.localizedDescription];
                [self.plugin fireEvent:@"document" event:@"on.appopen.failed.load" withData:jsonStr];
                return;
            }

            self.appOpenAd = ad;
            self.loadTime = [NSDate date];
            self.appOpenAd.fullScreenContentDelegate = self;

            __weak typeof(self) weakSelf = self;
            self.appOpenAd.paidEventHandler = ^(GADAdValue * _Nonnull value) {
                NSString *jsonStr = [NSString stringWithFormat:@"{\"value\":%lld, \"currency\":\"%@\", \"precision\":%ld}",
                                     value.value.longLongValue, value.currencyCode, (long)value.precision];
                [weakSelf.plugin fireEvent:@"document" event:@"on.appopen.revenue" withData:jsonStr];
            };

            [self.plugin fireEvent:@"document" event:@"on.appopen.loaded" withData:nil];

            if (self.isAutoShow) {
                [self showAdIfAvailable];
            }
        }];
    });
}

- (void)showAppOpenAd:(CDVInvokedUrlCommand *)command {
    dispatch_async(dispatch_get_main_queue(), ^{
        [self showAdIfAvailable];
        if (command) {
            CDVPluginResult* res = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
            [self.plugin.commandDelegate sendPluginResult:res callbackId:command.callbackId];
        }
    });
}

- (void)showAdIfAvailable {
    if (self.isShowingAd) return;

    if (![self isAdAvailable]) {
        NSString *jsonStr = @"{\"message\":\"Ad not ready or expired\"}";
        [self.plugin fireEvent:@"document" event:@"on.appopen.failed.show" withData:jsonStr];
        return;
    }

    if ([UIApplication sharedApplication].applicationState != UIApplicationStateActive) {
        NSString *jsonStr = @"{\"code\":\"APP_IN_BACKGROUND\", \"message\":\"Ad presentation blocked: Application is currently in the background or inactive.\"}";
        [self.plugin fireEvent:@"document" event:@"on.appopen.failed.show" withData:jsonStr];
        return;
    }

    self.isShowingAd = YES;
    [self.appOpenAd presentFromRootViewController:self.plugin.viewController];
}

#pragma mark - Expiration Logic (4 Hours Rule)

- (BOOL)wasLoadTimeLessThanNHoursAgo:(int)n {
    if (!self.loadTime) return NO;
    NSDate *now = [NSDate date];
    NSTimeInterval timeIntervalBetweenNowAndLoadTime = [now timeIntervalSinceDate:self.loadTime];
    double secondsPerHour = 3600.0;
    double intervalInHours = timeIntervalBetweenNowAndLoadTime / secondsPerHour;
    return intervalInHours < n;
}

- (BOOL)isAdAvailable {
    return self.appOpenAd != nil && [self wasLoadTimeLessThanNHoursAgo:4];
}

#pragma mark - GADFullScreenContentDelegate

- (void)adDidRecordImpression:(id<GADFullScreenPresentingAd>)ad {
    [self.plugin fireEvent:@"document" event:@"on.appopen.impression" withData:nil];
}

- (void)adDidRecordClick:(id<GADFullScreenPresentingAd>)ad {
    [self.plugin fireEvent:@"document" event:@"on.appopen.clicked" withData:nil];
}

- (void)adWillPresentFullScreenContent:(id<GADFullScreenPresentingAd>)ad {
    self.isShowingAd = YES;
    [self.plugin fireEvent:@"document" event:@"on.appopen.shown" withData:nil];
}

- (void)adDidDismissFullScreenContent:(id<GADFullScreenPresentingAd>)ad {
    self.appOpenAd = nil;
    self.isShowingAd = NO;
    [self.plugin fireEvent:@"document" event:@"on.appopen.dismissed" withData:nil];
}

- (void)ad:(id<GADFullScreenPresentingAd>)ad didFailToPresentFullScreenContentWithError:(NSError *)error {
    self.appOpenAd = nil;
    self.isShowingAd = NO;

    NSString *jsonStr = [NSString stringWithFormat:@"{\"code\":%ld, \"message\":\"%@\"}", (long)error.code, error.localizedDescription];
    [self.plugin fireEvent:@"document" event:@"on.appopen.failed.show" withData:jsonStr];
}

@end
